best-first-cv/script.whizzml

```diff
@@ -0,0 +1,561 @@
+                                             candidates
+                                             potentials)
+              _ (log-info "Next selected feature: " (next-feat-info "feature")))
+          (recur (append selected (next-feat-info "feature"))
+                 (make-complementary potentials [(next-feat-info "feature")])
+                 (cons {"features" (append selected
+                                           (next-feat-info "feature"))
+                              "phi" (next-feat-info "phi")} queue)))))))
+
+;; A simple function to get the max value in a list
+(define (get-max xs) (reduce (lambda (x y) (if (> x y) x y)) (head xs) xs))
+
+
+;; Filters the features in one list to be complementary with the ones in
+;; another list
+(define (make-complementary y-list x-list)
+  (iterate (acc y-list x x-list)
+    (filter (lambda (y) (not (= x (y "feature")))) acc)))
+
+
+;; Clean the created datasets and deliver a human-readable response
+(define (response dataset-id k-fold-datasets selected queue)
+  (map safe-delete k-fold-datasets)
+  {"selected-fields" (feature-names dataset-id selected)
+   "iterations-info" queue})
+
+
+;; Get feature names given ids
+(define (feature-names dataset-id ids)
+  (let (fields ((fetch dataset-id) "fields"))
+    (map (lambda (id) (fields [id "name"])) ids)))
+
+;; Get the default set of input fields for this dataset (all preferred
+;; fields minus the objective field).
+(define (default-inputs dataset-id obj-id)
+  (let (fields ((fetch dataset-id) "fields")
+        fids (keys fields))
+    (filter (lambda (k) (and (fields [k "preferred"] false) (not (= obj-id k))))
+            fids)))
+
+
+;; Get the set of input fields for this dataset by using the importance of the
+;; fields in a random decision forest
+(define (important-fields dataset-id obj-id)
+  (let (ensemble-id (create-and-wait-ensemble dataset-id
+                                              {"randomize" true
+                                               "objective_field" obj-id})
+        ensemble (fetch ensemble-id)
+        importance (ensemble "importance")
+        importance (for (key (keys importance))
+                     {"feature" key "phi" (importance key)}))
+    (reverse (sort-by-key "phi" importance))))
+
+
+;; Make a list of input fields candidates given the potential
+;; features in potentials.
+(define (make-candidates selected potentials)
+  (let (make-req (lambda (fid)
+                   (cons fid selected)))
+    (map make-req potentials)))
+
+;; Given a set of dataset folds, a list of potential features, and a list of
+;; model ids corresponding to those features, select the best
+;; potential feature by performing an evaluation on each model and
+;; returning the feature with the best performance.
+(define (select-feature k-fold-datasets objective-id options candidates potentials)
+  (let (make-attrs (lambda (inp) (merge {"input_fields" inp}
+                                        options))
```

Comment:
just double checking: here you're using merge instead of assoc because you want "input_fields" in `options` (if any) to take precedence over `inp`?

Reply:
hmm, they should because they are the potential fields to be selected by the algorithm, but I don't think that using assoc would work differently, would it?